main/src/ca/uwaterloo/flix/runtime/channel/channel.flix

```diff
@@ -0,0 +1,75 @@
+def newId(): ##GlobalCounter & Impure = 
+    import new flix.runtime.GlobalCounter() as newGlobalCounter;
+    import flix.runtime.GlobalCounter:newId() as javaNewId;
+    javaNewId(newGlobalCounter())
+
+type alias Lock = ##java.util.concurrent.locks.Lock
+
+type alias ReentrantLock = ##java.util.concurrent.locks.ReentrantLock
+
+def newReentrantLock(): ReentrantLock & Impure = 
+    import new java.util.concurrent.locks.ReentrantLock() as javaNewReentrantLock;
+    javaNewReentrantLock()
+
+type alias Condition = ##java.util.concurrent.locks.Condition
+
+def newCondition(reentrantLock: ReentrantLock): Condition & Impure =
+    import java.util.concurrent.locks.ReentrantLock.newCondition() as javaNewCondition;
+    javaNewCondition(reentrantLock)
+
+type alias LockConditionPair = {
```

Comment:
Records are a tad more "expensive" than tuples, so I would just go for the pair `(Lock, Condition)`.

Reply:
Should we always use tuples instead of records? For instance, should we also change `Channel` to be a tuple?

Reply:
I think I would recommend that. 

The very brief tldr is: Records have O(n) access time in the number of fields. Its a consequence of supporting extensible records. This is normally not a problem, except you are writing performance critical code.